File: milbot/plugins/exit.py
import re
import slack
import sys

from log import Log
import utils


@slack.RTMClient.run_on(event="message")
async def bot_exit(**payload):
    """bot を終了させる"""

    data = payload["data"]
    if utils.is_bot_message(data):
        return
    web_client = payload["web_client"]
    channel_id = data.get("channel")
    text = data.get("text")

    if re.match(r"^milbot exit help", text, re.IGNORECASE):
        try:
            await web_client.chat_postMessage(
                channel=channel_id,
                text="milbot を終了するコマンドです(｀･ω･´)"
            )
        except Exception as e:
            Log.error(e)
            raise(e)

    elif re.match(r"^milbot exit", text, re.IGNORECASE):
        try:
            web_client.chat_postMessage(
                channel=channel_id,
                text="milbot を終了します(｀-ω-´)zzZ"
            )
        finally:
            sys.exit()

File: botplugins/exit/exit.go
package exit

import (
	"context"
	"fmt"
	"log"
	"os"
	"regexp"

	"github.com/high-moctane/milbot/botlog"
	"github.com/slack-go/slack"
)

// validPrefix は反応するメッセージの正規表現です。
var validRegexp = regexp.MustCompile(`(?i)^milbot exit`)

// Plugin は終了コマンドを受け付けるプラグインです
type Plugin struct {
	client *slack.Client
}

// New でプラグインを生成します。
func New() *Plugin {
	return new(Plugin)
}

// Start でプラグインを有効化します。
func (p *Plugin) Start(client *slack.Client) error {
	p.client = client
	return nil
}

// Serve で終了コマンドを受け付けて終了します。
func (p *Plugin) Serve(ctx context.Context, event slack.RTMEvent) error {
	if !p.isValidEvent(event) {
		return nil
	}

	defer os.Exit(0)

	ev, _ := event.Data.(*slack.MessageEvent)
	user, err := p.getUserNameContext(ctx, ev)
	if err != nil {
		return fmt.Errorf("exit failed: %w", err)
	}

	_, _, _, err = p.client.SendMessageContext(
		ctx,
		ev.Channel,
		slack.MsgOptionText("Bye (｀･ω･´)", true),
	)
	if err != nil {
		return fmt.Errorf("exit failed: %v", err)
	}
	log.Printf("received exit command by %s", user)
	botlog.SendfContext(ctx, "received exit command by %s", user)
	return nil
}

// getUserName は event の送り主の名前を取得します。
func (p *Plugin) getUserNameContext(ctx context.Context, event *slack.MessageEvent) (name string, err error) {
	user, err := p.client.GetUserInfoContext(ctx, event.User)
	if err != nil {
		err = fmt.Errorf("could not get user: %w", err)
		return
	}
	name = user.Name
	return
}

// isValidEvent は event に反応するべきかどうか返します。
func (*Plugin) isValidEvent(event slack.RTMEvent) bool {
	ev, ok := event.Data.(*slack.MessageEvent)
	if !ok {
		return false
	}
	return validRegexp.MatchString(ev.Text)
}

// Stop でプラグインの終了処理をします。
func (p *Plugin) Stop() error {
	return nil
}

// Help でヘルプメッセージを返します。
func (p *Plugin) Help() string {
	return "[Exit]\n" +
		"`milbot exit` を受け取って bot を終了します。"
}
